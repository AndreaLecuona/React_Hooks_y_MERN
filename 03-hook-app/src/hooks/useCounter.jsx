import { useState } from "react"


//centraliza la lógica de "contador" para exportarla y reutilizarla en todos los componentes que la necesiten (y asi evitar repetir codigo)
export const useCounter = ( initialState = 1 ) => {

    const [counter, setCounter] = useState(initialState);

    const increment = () => {
        setCounter( counter + 1 );
    }

    const decrement = () => {
        setCounter( counter - 1 ); 
    }

    const reset = () => {
        setCounter( initialState );
    };

    //flexibilizo aun más, dandole la posibilidad de modificar en cuánto debe crecer o decrecer
    // const increment = ( factor = 1) => {
    //     setCounter( counter + factor );
    // }

    // const decrement = ( factor = 1) => {
    //     setCounter( counter - factor ); 
    // }

    // const reset = () => {
    //     setCounter( initialState );
    // };

    return{
        counter,
        increment,
        decrement,
        reset
    };
}
