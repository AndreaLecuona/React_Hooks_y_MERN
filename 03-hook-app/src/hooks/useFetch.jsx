import { useEffect, useRef, useState } from "react";

//custom hook que hace fetch de los datos de un endpoint que se le pasa como argumento
export const useFetch = (url) => {


    //USEREF para que este custom hook sea más seguro y dé menos lugar a errores
    //hacemos que isMounted (el useRef) mantenga la referencia siempre que este hook esté vivo o cuando el componente que lo usa siga montado.
    //cuando cambiemos los valores del isMounted, no queremos que vuelva a renderizarse el componente
    const isMounted = useRef(true);

    useEffect(() => {
        //no tiene efecto, solo aprovechamos la limpieza

        return () => {
            isMounted.current = false;
        }
    }, []); //se ejecuta una sola vez onpage load



    const [state, setState] = useState({ data: null, loading: true, error: null });

    useEffect(() => {

        setState({ data: null, loading: true, error: null }); //para que vuelva a aparecer el loading entre fetchs de frases

        fetch(url)
            .then(resp => resp.json())
            .then(data => {

                //si el componente esta montado DE MANERA SEGURA, entonces continuamos la ejecución que devuelve la data. Si no, no hacemos nada y evitamos un error que crashea el programa
                if( isMounted.current ){

                    setState({
                        loading: false,
                        error: null,
                        data
                    })

                } else {
                    console.log('setState no se ejecutó');
                }


            })
        .catch( () => {
            setState({
                data: null,
                loading: false,
                error: 'No pudo obtenerse la data'
            })
        })

    }, [url]);

    return state;

};
