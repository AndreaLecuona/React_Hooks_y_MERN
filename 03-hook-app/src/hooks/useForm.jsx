import { useState } from "react";

//extraemos la logica de las forms, que de otra forma se repetiría a lo largo de las secciones de una página
export const useForm = ( initialState = {} ) => {

    const [values, setValues] = useState(initialState);

    const reset = () => {
        setValues( initialState );
    }

    const handleInputChange = ( { target } ) => {

        setValues( {
            ...values,
            [ target.name ]: target.value
        });
    };

    return [ values, handleInputChange, reset ];

};
