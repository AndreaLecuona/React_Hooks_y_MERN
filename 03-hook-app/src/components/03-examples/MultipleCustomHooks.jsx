import React from 'react';
import { useCounter } from '../../hooks/useCounter';
import { useFetch } from '../../hooks/useFetch';
import '../02-useEffect/effects.css';

//Vamos a relacionar varios custom hooks
export const MultipleCustomHooks = () => {

    const { counter, increment } = useCounter(1);


    //const state = useFetch( `https://www.breakingbadapi.com/api/quotes/1` );
    const { loading, data } = useFetch(`https://www.breakingbadapi.com/api/quotes/${ counter }`);
    
    const { author, quote } = !!data && data[0]; //si hay data la extraigo, evitamos el error de extraerla cuando todavia está en null
                                                //la doble negacion está para asegurarnos que si es null, éste se traduzca a false y no ejecute nada (y por ende no rompa el codigo usando null como dato) --->  !null = true, pero !!null = false (no es null otra vez)


    return (
        <div>
            <h1>Breaking Bad quotes</h1>
            <hr />

            {/* ternario (podría simplificarse creando dos componentes con el div y el blockquote) */}
            {loading ?
                <div className="alert alert-info text-center">
                    Loading...
                </div>
                :
                <blockquote className="blockquote">
                    <p> { quote } </p>
                    <footer className="blockquote-footer"> { author } </footer>
                </blockquote>
            }

            <button className="btn btn-primary" onClick={ increment }>
                Siguiente quote
            </button>

        </div>
    )
}
