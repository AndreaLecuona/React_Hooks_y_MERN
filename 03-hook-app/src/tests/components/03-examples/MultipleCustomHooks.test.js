import React from "react";
import { shallow } from "enzyme";
import { MultipleCustomHooks } from "../../../components/03-examples/MultipleCustomHooks";
import { useFetch } from "../../../hooks/useFetch";
import { useCounter } from "../../../hooks/useCounter";

jest.mock( "../../../hooks/useFetch" );
jest.mock( "../../../hooks/useCounter" );



describe( 'Pruebas en MultipleCustomHooks', () => {


    beforeEach( () => {
        useCounter.mockReturnValue({
            counter: 10,
            increment: () => {}
        })
    })


    test('debe renderizarse correctamente', () => {

        //si no especificamos el resultado, el mock da error
        useFetch.mockReturnValue( {
            data: null,
            loading: true,
            error: null
        } );
        
        const wrapper = shallow( <MultipleCustomHooks/> );
        expect( wrapper ).toMatchSnapshot();

    })

    test('Debe mostrar la información cuando ya se llamó la API', () => {
        
        //como useFetch ya está testeado y no nos interesa evaluar su implementación sino sus resultados (data), hacemos un mock (simulación)
        // del use fetch y la respuesta (data)
        useFetch.mockReturnValue( {
            data: [{
                author: "Bowie",
                quote: "Let's dance"
            }],
            loading: false,
            error: null
        } );

        //capturamos el componente despues de la respuesta
        const wrapper = shallow( <MultipleCustomHooks/> );

        // console.log(wrapper.html());

        //evaluamos. Puede ser de varias formas: 
        // - (A) por snapshot
        // - (B) buscando clases que no existen ya que lo elementos que las tiene no se renderizan cuando hay data
        // - (C) buscando el texto que deberia estar

        expect( wrapper.find( '.alert' ).exists() ).toBe( false );  // (B)clase alert del loading no deberia existir

        expect( wrapper.find( 'p' ).text().trim() ).toBe( "Let's dance" ); // (C)texto resultado del mock
        expect( wrapper.find( '.blockquote-footer' ).text().trim() ).toBe( "Bowie" );
    })

});