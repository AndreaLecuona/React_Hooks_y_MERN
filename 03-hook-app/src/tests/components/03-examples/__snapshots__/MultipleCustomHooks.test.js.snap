// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en MultipleCustomHooks debe renderizarse correctamente 1`] = `
<div>
  <h1>
    Breaking Bad quotes
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Siguiente quote
  </button>
</div>
`;
